### ISSUE TITLE
[java] NullAssignment: False positive when assigning `null` to a `final` field in a constructor

### ISSUE THREAD

**Issue by Vampire** [pmd/pmd#6276]
**Affects PMD Version:**  7.18.0

**Rule:** [NullAssignment](https://docs.pmd-code.org/latest/pmd_rules_java_errorprone.html#nullassignment)

**Description:**

If you have a `final` field that might be `null` or not, depending on the called constructor or depending on the arguments given to the constructor, you **must** set it to `null` explicitly in the constructor.
If you set it at declaration time, you cannot set it to a different value.
So assigning `null` to a `final` field in a constructor should not trigger a finding.

This does sometimes work as expected for example when directly assigning `null`, but if it is done as part of a ternary expression for example, the rule complains.

**Code Sample demonstrating the issue:**

https://github.com/Vampire/command-framework/blob/4e8f3137014399bc3e0673077239dbd641b63b20/src/main/java/net/kautler/command/api/Version.java#L95

```java
import java.time.Instant;

public class Version {
    private final Instant buildTimestamp;

    public Version(String buildTimestamp) {
        this.buildTimestamp = "$buildTimestamp".equals(buildTimestamp) ? null : Instant.parse(buildTimestamp);
    }
}
```

**Expected outcome:**

PMD reports a violation at line 95, but that's wrong. That's a false positive.

**Running PMD through:** *Gradle*

**Comment by mrclmh** [comment]
Extracted example:

```Java
import java.time.Instant;

public class Version {
    private final Instant buildTimestamp;

    public Version() {
        String buildTimestamp = "1764222289";
        this.buildTimestamp = "$buildTimestamp".equals(buildTimestamp) ? null : Instant.parse(buildTimestamp);
    }
}

```

**Comment by mrclmh** [comment]
Another ternary problem mentioned here https://github.com/pmd/pmd/issues/1050#issuecomment-410246566.
